packages/gasket-plugin-nextjs/lib/next-route.js

```diff
@@ -0,0 +1,81 @@
+/* eslint-disable spaced-comment */
+// @ts-check
+/// <reference types="@gasket/engine" />
+
+const path = require('path');
+const fs = require('fs').promises;
+
+/**
+ * @typedef {import("@gasket/engine").Gasket} Gasket
+ * @typedef {import("http").IncomingMessage}  Request
+ */
+
+// Because getNextRoute may be called multiple times during a request, cache
+const nextRouteForRequest = new WeakMap();
+
+/**
+ * Gets the NextJS route matching the request
+ *
+ * @param {Gasket}  gasket  The gasket engine
+ * @param {Request} req     The HTTP request
+ *
+ * @returns {Promise<Object | null>} A Next.JS route object or null
+ */
+async function getNextRoute(gasket, req) {
+  if (nextRouteForRequest.has(req)) {
+    return nextRouteForRequest.get(req);
+  }
+
+  let result = null;
+  const routes = await loadRoutes(gasket);
+  if (routes) {
+    for (const route of iterateRoutes(routes)) {
+      if (route.regex.test(req.url)) {
+        result = route;
+        break;
+      }
+    }
+  }
+
+  nextRouteForRequest.set(req, result);
+  return result;
+}
+
+let routesPromise, routesLoadError, cachedRoutes;
+async function loadRoutes(gasket) {
+  if (routesLoadError) {
+    return null;
+  }
+
+  if (cachedRoutes) {
+    return cachedRoutes;
+  }
+
+  if (!routesPromise) {
+    routesPromise = fs
+      .readFile(path.join(gasket.config.root, './.next/routes-manifest.json'))
```

Comment:
There doesn't seem to be any sort of official support within next.js to tap into its router that I could find. That would be preferable to reading a manifest JSON file which we don't necessarily have guarantees on for format, but it's the cleanest I could come up with so far. I wish they supported more abilities for custom servers, but I'm guessing they're not highly motivated to do so since they sell their own hosting platform.

Reply:
Do you need to let `path.join` handle the separators to keep things safe for windows?
```suggestion
      .readFile(path.join(gasket.config.root, '.next', 'routes-manifest.json'))
```

Reply:
Node.js normalizes things. Tried this on Windows:

```
> require('path').join('C:\\Windows', '.next/routes-manifest.json')
'C:\\Windows\\.next\\routes-manifest.json'
```

Reply:
Updated it anyway